### PR TITLE
[dynamo,optim] Place guards on the args before assuming they exist

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118630
* #118855
* #118535
* #118199
* #118208
* #118003
* #118194
* #117625
* __->__ #117983
* #118098
* #117982

This enables the new way of writing guards for dicts. Before we were
doing things like
```
  L['self'].param_groups[0][___dict_keys_getitem(L['self'].param_groups[0], 0)][3] is L['self'].param_groups[0]['params'][3]
```
without knowing whether `L['self'].param_groups[0][___dict_keys_getitem(L['self'].param_groups[0], 0)]` was a list.

On a different note, I'll probably write a pass to recover the previous
way to place guards on dicts via something like `DICT_KEYS`  as an
optimisation, as it seems relevant for optimisers.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/optimizer.py [hunk 1]
```diff
@@ -60,6 +60,14 @@ def call_method(
                 ret_val = self.value._init_group(*py_args, **py_kwargs)
                 self.map_sources_and_install_guards(tx)
                 self.update_list_args(tx, args, kwargs, py_args, py_kwargs)
+                # this next line has the side effect of installing guards
```
Comment (mlazos) [inline torch/_dynamo/variables/optimizer.py:63]: Can you explain this comment a little more, I don't see why this needs to be the case. 

We install guards on all the param groups previously, why does this order matter?

Comment (lezcano) [reply torch/_dynamo/variables/optimizer.py:63]: See the OP. The example where this broke was
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_optim.py -k test_can_load_older_state_dict_ASGD_cpu_float32
```

The issue was that the internals of the optimizer had changed, so `___dict_keys_getitem(d, 1)` was now the learning rate rather than `params`, and we were installing first guards of the form
`___dict_keys_getitem(d, 1)[0] is <some other tensor>` before checking that `___dict_keys_getitem(d, 1)` was even a list

Comment (mlazos) [reply torch/_dynamo/variables/optimizer.py:63]: We would have to construct the list during recursive_realize though right?

Comment (lezcano) [reply torch/_dynamo/variables/optimizer.py:63]: So this is a bit of a corner case bug of https://github.com/pytorch/pytorch/pull/117625. In
```
                    GetItemSource(group_source, "params"), p_ind
```
we define the source of some parameters using the string `"params"`, while when tracing a dict, we do so by using the more verbose `___dict_keys_getitem(L['self'].param_groups[0], 5)` (this expression evaluates to `"params"`). Then, when calling `VariableBuilder` on `param_groups`, it looks at the two sources for these tensors and it thinks that they are different, so it eagerly places an unnecessary `make_dupe_guard`. This guard is put there before the other guards on the types are put in place, so we hit an error.

I'll add a PR into this stack that removes this guard entirely, and I will revert these changes then.

Comment (lezcano) [reply torch/_dynamo/variables/optimizer.py:65]: Added a patch fixing these unnecessary guards (and pretty much reverting this PR) at https://github.com/pytorch/pytorch/pull/118535.

**Review by mlazos — APPROVED** [review]
